src/native/shared/perfetto_support.hh

```diff
@@ -0,0 +1,195 @@
+#if !defined(PERFETTO_SUPPORT_HH)
+#define PERFETTO_SUPPORT_HH
+
+#if defined(PERFETTO_ENABLED)
+#include <string_view>
+#include <type_traits>
+
+#include <mono/jit/jit.h>
+#include <mono/metadata/appdomain.h>
+#include <mono/metadata/assembly.h>
+#include <mono/metadata/class.h>
+#include <mono/metadata/debug-helpers.h>
+#include <mono/metadata/mono-config.h>
+#include <mono/metadata/mono-debug.h>
+#include <mono/metadata/object.h>
+#include <mono/utils/mono-dl-fallback.h>
+#include <mono/utils/mono-logger.h>
+
+#include <perfetto.h>
+
+#include "cpp-util.hh"
+
+namespace xamarin::android {
+	class PerfettoConstants
+	{
+	public:
+		static constexpr std::string_view MonoRuntimeCategory         { "mono-runtime" };
+		static constexpr std::string_view MonodroidCategory           { "monodroid" };
+
+		static constexpr std::string_view AssemblyLoadAnnotation      { "Assembly load" };
+		static constexpr std::string_view ImageLoadAnnotation         { "Assembly image load" };
+		static constexpr std::string_view ClassLoadAnnotation         { "Class load" };
+		static constexpr std::string_view VTableLoadAnnotation        { "VTable load" };
+		static constexpr std::string_view MethodInvokeAnnotation      { "Method: invoke" };
+		static constexpr std::string_view MethodRunTimeAnnotation     { "Method: inner run time" };
+		static constexpr std::string_view MonitorContentionAnnotation { "Monitor contention" };
+		static constexpr std::string_view MonodroidRuntimeTrack       { "Monodroid" };
+
+		static constexpr std::string_view XAInitInternal              { "InitInternal" };
+	};
+
+	enum class PerfettoTrackId : uint64_t
+	{
+		// We need to start high, so that we don't conflict with the standard Perfetto trakck IDs
+		AssemblyLoadMonoVM = 0xDEADBEEF,
+		ClassLoadMonoVM,
+		ImageLoadMonoVM,
+		MethodInnerMonoVM,
+		MethodInvokeMonoVM,
+		MonitorContentionMonoVM,
+		VTableLoadMonoVM,
+
+		MonodroidRuntime,
+	};
+}
+
+PERFETTO_DEFINE_CATEGORIES (
+	perfetto::Category (xamarin::android::PerfettoConstants::MonoRuntimeCategory.data ()).SetDescription ("Events from the MonoVM runtime"),
+	perfetto::Category (xamarin::android::PerfettoConstants::MonodroidCategory.data ()).SetDescription ("Events from the .NET Android native runtime")
+);
+
+namespace xamarin::android {
+	namespace detail {
+		template<typename TMonoType>
+		concept SupportedMonoType = requires {
```

Comment:
What odd syntax that `requires { requires` is a valid sequence of tokens.  Wat?

Reply:
It's not strictly required in this case, I just found it more expressive, but maybe the repetition does look weird.  It represents a "nested requirement" which may make the concept a bit easier to read, instead of listing the requirements in the "main" constraint expression (which may take a form of, e.g., `requires (TMonoType t, size_t some_size)`), making it look less crammed.  I'll remove the extra `requires` from here, since it isn't strictly necessary.